File: easy/reverse_linked_list/reverse_linked_list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
        return

    def push(self, val):
        new_node = Node(val)
        curr = self
        while curr.next != None:
            curr = curr.next
        curr.next = new_node
        return

    def print(self):
        curr = self
        while curr != None:
            print(curr.val, end=" -> ")
            curr = curr.next
        print()


class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None or head.next == None:
            return head
        prev, curr = None, head
        while curr.next != None:
            curr.next, prev, curr = prev, curr, curr.next
        curr.next = prev
        return curr


s = Solution()

print("Solution 1")
head = Node(1)
head.push(2)
head.push(3)
head.push(4)
head.push(5)
head.print()
reverse = s.reverseList(head)
reverse.print()

File: easy/linked_list/reverse_linked_list/reverse_linked_list.go
package main

import "fmt"

type Node struct {
	Val  int
	Next *Node
}

func (n *Node) push(val int) {
	newNode := &Node{Val: val}
	curr := n
	for curr.Next != nil {
		curr = curr.Next
	}
	curr.Next = newNode
	return
}

func (n *Node) print() {
	curr := n
	for curr != nil {
		fmt.Printf(" %d -> ", curr.Val)
		curr = curr.Next
	}
	fmt.Println()
	return
}

func reverseList(head *Node) *Node {
	if head == nil || head.Next == nil {
		return head
	}

	prev, curr := head, head.Next
	head.Next = nil

	for curr.Next != nil {
		curr.Next, prev, curr = prev, curr, curr.Next
	}
	curr.Next = prev

	return curr
}

func main() {
	fmt.Println("Solution 1")
	head := &Node{Val: 1}
	head.push(2)
	head.push(3)
	head.push(4)
	head.push(5)
	head.print()
	reverse := reverseList(head)
	reverse.print()
}
